### PR TITLE
perf: use new PluginDriver cache

### PR DESCRIPTION
**Pull request by keithamus**
### What?

This is a pre-emptive PR on https://github.com/rollup/rollup/pull/2382 - to help guide the implementation of `PluginDriver.cache`.

### Why?

The intent is to greatly speed up how this plugin works, and also give an idea of how Rollup core might implement a cache layer.

This will speed up resolving because as opposed to reading and stating multiple files, it should reduce the load to just stating files it needs to, and returning the same result.

### How?

This checks to see if `this.cache` exists, and if it does it will attempt to read from that cache. A successful read of the cache is then checked for validity, and returned instead of executing the entire resolve algorithm. If the cache is not present or invalid, then the algorithm is run as normal.

What consitutes a valid cache? Every file that was read as part of the algorithm (this includes files that don't actually exist, but the algorithm needs to check for existence) is stated and the `mtimeMs` and `size` are stored. The validation routine then checks the _cached_ values of those, against a live stat call and compares the result. If any result is different the cache is blown away. In other words if the user changes or adds a `package.json` or `index.js` file, or changes the importee, importer, then the cached result is disregarded as it is assumed unsafe.

It still has to `stat` a lot of files to ensure the cache is valid, but on the other hand, the number of files read is reduced to 0 on a hot cache. This should make a significant difference to resolve times. Even if the cache is marked as invalid, the work is not wasted as all stat calls get cached for the duration of the build and so subsequent stat calls happen with no IO.

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -86,116 +108,146 @@ export default function nodeResolve ( options = {} ) {
...
-								if ( !extname(key) ) {
-									extensions.reduce( ( browser, ext ) => {
-										browser[ absoluteKey + ext ] = browser[ key ];
-										return browser;
-									}, browser );
-								}
-							}
-							return browser;
-						}, {});
+			return Promise.resolve(this.cache ? this.cache.get(`${importee}|${importer}`) : false)
```
Comment (guybedford) [inline src/index.js]: I wonder if we should have cache loading async, or if we can just assume that the initial async cache load has already completed so that the cache is always synchronous.

The only time this might be an issue is if the cache grows very large, and we want to load the cache itself partially with an index or something. But on a per-project level I struggle to see us getting there? I mean loading a few MB into memory shouldn't be an issue right?

File: src/index.js [hunk 2]
```diff
@@ -86,116 +108,146 @@ export default function nodeResolve ( options = {} ) {
...
+				.then((cache) => {
+					// If the importer, importee or any package.json files have changed then invalidate the cache:
+					return cache && cache.stats.reduce((promise, cacheEntry) => {
+						return promise.then(stillValid => stillValid && compareFileCacheEntry(cacheEntry), Promise.resolve(true));
```
Comment (guybedford) [inline src/index.js]: So the gap here, is that a file that isn't tracked by the cache can in theory invalidated the resolver cache. Although perhaps this is enough of an edge case not to be an issue.

Example scenarios would be adding `file.mjs` when `file.js` was loaded previously.

File: src/index.js (hunk 1 above)
Comment (keithamus) [reply src/index.js]: I had considered this - but even for some really overblown cache we're still going to be looking at sub 100mb cache sizes. This plugin will to store around ~100bytes per entry on average for maybe 1000-2000 entries for a larger build. So just this plugin will maybe store 10-20kb of cache. The transform cache could be larger - perhaps in the megabytes as it has to store input code and output code; but cache in the megabytes will still be quick enough to load from disk before any plugins are called.

File: src/index.js (hunk 2 above)
Comment (keithamus) [reply src/index.js]: My understanding is that just adding `file.mjs` would not change how the module is resolved - unless other files around it changed. Is this correct?

The `cache.stats` list will store the importer, importee stats - plus any possible `package.json` files that could influence them, plus any `index.js` files it could potentially hit. In other words adding files which might intercept the loader algorithm would invalidate the cache precisely because we check every file that could intercept the loader algorithm.

File: src/index.js (hunk 1 above)
Comment (guybedford) [reply src/index.js]: Ok agreed. In that case, I think we can assume `this.cache` is available synchronously for plugins.

File: src/index.js (hunk 2 above)
Comment (guybedford) [reply src/index.js]: Node.js will resolve `file | file.mjs | file.js | file.json | file.node`. So if you're loading one of the later ones, then adding an earlier entry will take precedence regardless of the stat of the previous file.

Similarly it will resolve `dir/package.json>main | dir/index.mjs | dir/index.js | dir/index.json | dir/index.node` so that a similar precedence applies there.

Then there is also the case of `node_modules/pkgX` checking `node_modules/pkgX/node_modules/pkgY` before `node_modules/pkgY` so that if we have cached the lookup there to `node_modules/pkgY` then adding a `node_modules/pkgX/node_modules/pkgY` won't be seen.

For these reasons the cache approach I'd recommend is always re-running the resolver, but with a stat cache and package.json cache with a timed invalidation (at the risk of repeating myself, but admittedly I know I've skipped the above logic until now).

Comment (keithamus) [reply src/index.js]: I ran a quick test with the `node-resolve` function to see what files it looked for. The importer was `path/to/foo.js` which imports `fetch` (which is found at `node_modules/fetch/fetch.js`). `node-resolve` will look for the following files:
 - `path/to/node_modules/fetch`
 - `path/to/node_modules/fetch.js`
 - `path/to/node_modules/fetch/package.json`
 - `path/to/node_modules/fetch/index`
 - `path/to/node_modules/fetch/index.js`
 - `path/to/fetch`
 - `path/to/fetch.js`
 - `path/to/fetch/package.json`
 - `path/to/fetch/index`
 - `path/to/fetch/index.js`
 - `path/node_modules/fetch`
 - `path/node_modules/fetch.js`
 - `path/node_modules/fetch/package.json`
 - `path/node_modules/fetch/index`
 - `path/node_modules/fetch/index.js`
 - `path/fetch`
 - `path/fetch.js`
 - `path/fetch/package.json`
 - `path/fetch/index`
 - `path/fetch/index.js`
 - `node_modules/fetch`
 - `node_modules/fetch.js`
 - `node_modules/fetch/package.json` (found this file!)
 - `node_modules/fetch/fetch.js` (found this file!)

`node-resolves` checks every file with `isFile()` and `readFile()` which we instrument. Every one of those files gets statted as part of this code. Only 2 out of the 24 files exist. The ones that don't are still stored in the cache as `mtimeMs: 0, size: 0`. So if I later go and add `node_modules/fetch.js` to intercept that path, then the stat cache for that file will not match and the whole cache is invalidated.

In other words, any file `node-resolve` attempts to read - we know about, and we store some info about, to later check.

Comment (guybedford) [reply src/index.js]: That definitely catches everything then, but my worry in this scenario is we are now storing all the stats per individual resolution, and then restatting them for each resolution right?

So we now have a lot of cached data to load and work with, as well as still doing all the stats. So where is the performance benefit then? Have you run the numbers on this?

**Comment by shellscape** [comment]
@keithamus is this one you'd like to see merged? We're getting ready to move this plugin to https://github.com/rollup/plugins and are tying up loose ends before that happens.

**Comment by keithamus** [comment]
I'm not sure what the situation is with the PluginCache currently; I'll need to spend time time investigating it all. We can close it for now and I'll renew it at a later time.

**Comment by shellscape** [comment]
Thank you for following up.
